File: docs/breez-sdk/snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/CrossChain.kt
package com.example.kotlinmpplib

import breez_sdk_spark.*
import com.ionspin.kotlin.bignum.integer.BigInteger

class CrossChain {
    suspend fun getCrossChainRoutes(sdk: BreezSdk) {
        // ANCHOR: cross-chain-get-routes
        val input = "<recipient address>"

        try {
            val parsed = sdk.parse(input)
            if (parsed !is InputType.CrossChainAddress) {
                throw IllegalArgumentException("Not a cross-chain address")
            }
            val addressDetails = parsed.v1

            val routes = sdk.getCrossChainRoutes(
                CrossChainRouteFilter.Send(addressDetails = addressDetails)
            )

            for (route in routes) {
                // Log.v("Breez", "Route via ${route.provider}: ${route.chain}/${route.asset}")
                // Amount bounds are published per accepted asset. Read the ones for
                // the asset you intend to pay with, before quoting.
                for (accepted in route.acceptedAssets) {
                    val limits = accepted.limits ?: continue
                    // Log.v("Breez", "  ${accepted.asset} minimum: ${limits.minAmount} " +
                    //     "base units / ${limits.minUsdCents} USD cents")
                    if (limits.dynamicLimitsPossible) {
                        // Log.v("Breez", "  The provider may enforce a higher minimum")
                    }
                }
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: cross-chain-get-routes
    }

    suspend fun prepareSendPaymentCrossChain(
        sdk: BreezSdk,
        addressDetails: CrossChainAddressDetails,
        route: CrossChainRoutePair,
    ) {
        // ANCHOR: cross-chain-prepare
        // Optionally set the maximum slippage in basis points (10 to 500)
        val optionalMaxSlippageBps: UInt? = 100u

        try {
            val req = PrepareSendPaymentRequest(
                paymentRequest = PaymentRequest.CrossChain(
                    address = addressDetails.address,
                    route = route,
                    maxSlippageBps = optionalMaxSlippageBps,
                    targetOverpayBps = null,
                ),
                amount = BigInteger.fromLong(50_000L),
                tokenIdentifier = null,
                conversionOptions = null,
                feePolicy = null,
            )
            val prepareResponse = sdk.prepareSendPayment(req)

            val paymentMethod = prepareResponse.paymentMethod
            if (paymentMethod is SendPaymentMethod.CrossChainAddress) {
                val amountIn = paymentMethod.amountIn
                val estimatedOut = paymentMethod.estimatedOut
                val feeAmount = paymentMethod.feeAmount
                val expiresAt = paymentMethod.expiresAt
                // Log.v("Breez", "Amount in: $amountIn")
                // Log.v("Breez", "Estimated out: $estimatedOut")
                // Log.v("Breez", "Provider fee: $feeAmount")
                // Log.v("Breez", "Quote expires at: $expiresAt")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: cross-chain-prepare
    }

    suspend fun sendPaymentCrossChain(
        sdk: BreezSdk,
        prepareResponse: PrepareSendPaymentResponse,
    ) {
        // ANCHOR: cross-chain-send
        // Only valid for sends with no token leg (see Retry safety).
        val optionalIdempotencyKey = "<idempotency key uuid>"
        try {
            val req = SendPaymentRequest(
                prepareResponse = prepareResponse,
                options = null,
                idempotencyKey = optionalIdempotencyKey,
            )
            val sendResponse = sdk.sendPayment(req)
            val payment = sendResponse.payment
            // Log.v("Breez", "Payment: $payment")
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: cross-chain-send
    }

    suspend fun getCrossChainReceiveRoutes(sdk: BreezSdk) {
        // ANCHOR: cross-chain-get-receive-routes
        try {
            val routes = sdk.getCrossChainRoutes(
                CrossChainRouteFilter.Receive(contractAddress = null)
            )

            for (route in routes) {
                println("Route via ${route.provider}: ${route.chain}/${route.asset} -> Spark")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: cross-chain-get-receive-routes
    }

    suspend fun receivePaymentCrossChain(sdk: BreezSdk, route: CrossChainRoutePair) {
        // ANCHOR: cross-chain-receive
        // amount is in the route's source-asset base units (USD-stable parity:
        // 1_000_000 = $1 on 6-decimal routes). See the guide for feeMode,
        // destination, and the slippage/overpay overrides.
        val amount = BigInteger.fromLong(1_000_000L)
        val optionalDestination: SparkAsset? = null
        val optionalMaxSlippageBps: UInt? = 100u
        val optionalTargetOverpayBps: UInt? = null
        val optionalFeeMode: CrossChainFeeMode? = null
        try {
            val req = ReceivePaymentRequest(
                paymentMethod = ReceivePaymentMethod.CrossChain(
                    route = route,
                    amount = amount,
                    destination = optionalDestination,
                    feeMode = optionalFeeMode,
                    maxSlippageBps = optionalMaxSlippageBps,
                    targetOverpayBps = optionalTargetOverpayBps,
                ),
            )
            val response = sdk.receivePayment(req)
            println("Payment request: ${response.paymentRequest}")
            val info = response.crossChainInfo
            if (info != null) {
                println("Deposit address: ${info.depositAddress}")
                println("Deposit amount: ${info.depositAmount}")
                println(
                    "Expected received: ${info.expectedReceivedAmount} " +
                        "${info.destinationAsset}",
                )
                println("Expires at: ${info.expiresAt}")
            }
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: cross-chain-receive
    }
}

File: docs/breez-sdk/snippets/wasm/cross_chain.ts
import {
  type BreezSdk,
  type CrossChainAddressDetails,
  type CrossChainRoutePair,
  type PrepareSendPaymentResponse
} from '@breeztech/breez-sdk-spark'

const exampleGetCrossChainRoutes = async (sdk: BreezSdk) => {
  // ANCHOR: cross-chain-get-routes
  const input = '<recipient address>'
  const parsed = await sdk.parse(input)
  if (parsed.type !== 'crossChainAddress') {
    throw new Error('Not a cross-chain address')
  }

  const routes = await sdk.getCrossChainRoutes({
    type: 'send',
    addressDetails: parsed
  })

  for (const route of routes) {
    console.debug(`Route via ${route.provider}: ${route.chain}/${route.asset}`)
    // Amount bounds are published per accepted asset. Read the ones for
    // the asset you intend to pay with, before quoting.
    for (const accepted of route.acceptedAssets) {
      const limits = accepted.limits
      if (limits === undefined) {
        continue
      }
      console.debug(
        `  ${accepted.asset.type} minimum: ${limits.minAmount} base units / ` +
          `${limits.minUsdCents} USD cents`
      )
      if (limits.dynamicLimitsPossible) {
        console.debug('  The provider may enforce a higher minimum than published')
      }
    }
  }
  // ANCHOR_END: cross-chain-get-routes
}

const examplePrepareSendPaymentCrossChain = async (
  sdk: BreezSdk,
  addressDetails: CrossChainAddressDetails,
  route: CrossChainRoutePair
) => {
  // ANCHOR: cross-chain-prepare
  // Optionally set the maximum slippage in basis points (10 to 500)
  const optionalMaxSlippageBps = 100

  const prepareResponse = await sdk.prepareSendPayment({
    paymentRequest: {
      type: 'crossChain',
      address: addressDetails.address,
      route,
      maxSlippageBps: optionalMaxSlippageBps
    },
    amount: BigInt(50_000),
    tokenIdentifier: undefined,
    conversionOptions: undefined,
    feePolicy: undefined
  })

  if (prepareResponse.paymentMethod.type === 'crossChainAddress') {
    const { amountIn, estimatedOut, feeAmount, expiresAt } = prepareResponse.paymentMethod
    console.debug(`Amount in: ${amountIn}`)
    console.debug(`Estimated out: ${estimatedOut}`)
    console.debug(`Provider fee: ${feeAmount}`)
    console.debug(`Quote expires at: ${expiresAt}`)
  }
  // ANCHOR_END: cross-chain-prepare
}

const exampleSendPaymentCrossChain = async (
  sdk: BreezSdk,
  prepareResponse: PrepareSendPaymentResponse
) => {
  // ANCHOR: cross-chain-send
  // Only valid for sends with no token leg (see Retry safety).
  const optionalIdempotencyKey = '<idempotency key uuid>'
  const sendResponse = await sdk.sendPayment({
    prepareResponse,
    options: undefined,
    idempotencyKey: optionalIdempotencyKey
  })
  const payment = sendResponse.payment
  console.debug('Payment:', payment)
  // ANCHOR_END: cross-chain-send
}

const exampleGetCrossChainReceiveRoutes = async (sdk: BreezSdk) => {
  // ANCHOR: cross-chain-get-receive-routes
  const routes = await sdk.getCrossChainRoutes({
    type: 'receive',
    contractAddress: undefined
  })

  for (const route of routes) {
    console.debug(
      `Route via ${route.provider}: ${route.chain}/${route.asset} -> Spark`
    )
  }
  // ANCHOR_END: cross-chain-get-receive-routes
}

const exampleReceivePaymentCrossChain = async (
  sdk: BreezSdk,
  route: CrossChainRoutePair
) => {
  // ANCHOR: cross-chain-receive
  // amount is in the route's source-asset base units (USD-stable parity:
  // 1_000_000 = $1 on 6-decimal routes). See the guide for feeMode,
  // destination, and the slippage/overpay overrides.
  const amount = '1000000'
  const optionalDestination = undefined
  const optionalMaxSlippageBps = 100
  const optionalTargetOverpayBps = undefined
  const optionalFeeMode = undefined

  const response = await sdk.receivePayment({
    paymentMethod: {
      type: 'crossChain',
      route,
      amount,
      destination: optionalDestination,
      feeMode: optionalFeeMode,
      maxSlippageBps: optionalMaxSlippageBps,
      targetOverpayBps: optionalTargetOverpayBps
    }
  })

  console.debug(`Payment request: ${response.paymentRequest}`)
  if (response.crossChainInfo !== undefined) {
    const {
      depositAddress,
      depositAmount,
      expectedReceivedAmount,
      destinationAsset,
      expiresAt
    } = response.crossChainInfo
    console.debug(`Deposit address: ${depositAddress}`)
    console.debug(`Deposit amount: ${depositAmount}`)
    console.debug(
      `Expected received: ${expectedReceivedAmount} ${destinationAsset}`
    )
    console.debug(`Expires at: ${expiresAt}`)
  }
  // ANCHOR_END: cross-chain-receive
}
